### ISSUE TITLE
Genome Browser & Table-data Browser: In the filter-criterion editor, there are errors regarding the coefficients for the "< Column" and "> Column" comparison operators

### ISSUE THREAD

**Issue by leehart** [wtchg-kwiatkowski/observatory-web#310]
1. Sometimes a critical error
(e.g. Samples table > add filter > add criterion > choose "Sampling year" and the "< Column" operator, and then try changing the linear and constant coefficients, boom); and

2. sometimes a funky non-critical error
(e.g. Samples table > add filter > add criterion > choose "Sampling year" and the "> Column" operator, and "Latitude", and then you should see that the default linear and constant coefficient values will change to "NULL", and cannot be changed, and that "NaN" appears in the SQL-esque representation)

I'm not sure if it's worth logging another issue for, but the option to add coefficients for these comparison operators was a surprise and a bit confusing at first, seeming like an afterthought bolted on to the otherwise straightforward "< Column" option. Not sure what to suggest to improve that though.

**Comment by leehart** [comment]
Re: (1) It looks like there's a bug or two in the `toQueryDisplayString` function of `SQL.WhereClause.CompareField` of Panoptes `SQL.js`, when it tries to get `queryData.fieldInfoMap[that.ColName2].name` but `that.ColName2` has *become* undefined.

**Comment by leehart** [comment]
Just to note: in the code, the linear coefficient is referred to as the "factor", and the constant coefficient is called the "offset".

**Comment by leehart** [comment]
Ah, I think I've cracked it: `handleValueChange` is designed to handle the change of one particular value, but it's being (trying to be) used to handle two input value changes (`scale` and `offset`) at the same time. (Makes a case for more formal specs for functions.)
So I'll separate those out and see...

**Comment by leehart** [comment]
Nope, there are still problems... I'm tempted to rewrite this whole feature!

**Comment by benjeffery** [comment]
Looks like a casualty from when mui changed fields over to being functional components that can't be `ref`fed.

**Comment by leehart** [comment]
Aha, one of the other problems was that `handleValueChange` was also trying to handle a potential `ColName2` change at the same time, which is handled separately.
I'm going to stop the running commentary now! :-)

**Comment by leehart** [comment]
Aw, ZenHub doesn't support thumbsups on comments. I'll have to do it GitHub...

Yes, I think you're right Ben: it looks like the new values aren't going through the onChange event. Thanks!

**Comment by leehart** [comment]
Just noticed there's a `handleValueEdit` handler as well, which the `minMax` uses for `onChange`, so that values aren't formatted while they're still being edited (which is one of the problems in this issue, so `scale` and `offset` should probably be using that instead), and `minMax` uses `handleValueChange` for `onBlur` to make the state change, and expects `handleValueChange` to update more than one value, so I either need to make sure `handleValueChange` can do that (properly), or amend those function calls.
In any case, I think I can see a way through this now.

**Comment by leehart** [comment]
I'm assuming that non-integer "factors" and "offsets" , e.g. 2.2, should be allowed.
(It looks like there's a related bug downstream that re-formats such values into integers, but doesn't display the altered values.)

**Comment by leehart** [comment]
ah, it's related to the type of the column being compared. (E.g. if the column is an integer, then the mechanism is formatting the "factors" and "offsets" as integers.)

**Comment by leehart** [comment]
There is also another "bug" (more of a user-interface issue) in this area, which I might have to log separately if I don't fix it while I have the box open: it seems intuitive to hit enter after specifying one of the values (factor or offset), but that has the unintended effect of closing the modal-dialog. This wouldn't be such a problem if the specified value was actually "saved", but because we're using onBlur to set the value after validation and formatting (while using onChange to update the value without validation and reformatting, while the user is typing) the user has to open the filter again and re-specify, being careful not to hit enter again.
A better behaviour would be to only have enter close-and-apply when focus was not in a form field, or perhaps suppress that behaviour when fields have focus.
